### DISCUSSION TITLE
Challenges compiling examples - compiler error or issues with my install of libusb?

### DISCUSSION THREAD

**Discussion by PCBscanner** [libusb/libusb#1570] [Q&A] [answered]
I'm facing challenges compiling the examples included with libusb. Here's the steps I have taken, and where I am up to.
Built libusb from source using `./autogen.sh`, `./configure`, `make`, `sudo make install`.
Installed to `/usr/local/include/libusb-1.0` - this is where libusb.h resides.

Trying to compile with `gcc testlibusb.c -o testlibusb` results in the following errors:

```
testlibusb.c:22:10: fatal error: libusb.h: No such file or directory
   22 | #include "libusb.h"
      |          ^~~~~~~~~~
compilation terminated.
```

Trying to compile by explicitly linking to the .h file: `gcc -I/usr/local/include/libusb-1.0 testlibusb.c -o testlibusb` results in the following errors. Note I do not see why I'd need to do this is there's a shared library, see below:
```

/usr/bin/ld: /tmp/ccL5eQxt.o: in function `print_endpoint':
testlibusb.c:(.text+0x161): undefined reference to `libusb_get_ss_endpoint_companion_descriptor'
/usr/bin/ld: testlibusb.c:(.text+0x184): undefined reference to `libusb_free_ss_endpoint_companion_descriptor'
/usr/bin/ld: /tmp/ccL5eQxt.o: in function `print_bos':
testlibusb.c:(.text+0x3f0): undefined reference to `libusb_get_bos_descriptor'
/usr/bin/ld: testlibusb.c:(.text+0x47d): undefined reference to `libusb_get_usb_2_0_extension_descriptor'
/usr/bin/ld: testlibusb.c:(.text+0x4a2): undefined reference to `libusb_free_usb_2_0_extension_descriptor'
/usr/bin/ld: testlibusb.c:(.text+0x4c5): undefined reference to `libusb_get_ss_usb_device_capability_descriptor'
/usr/bin/ld: testlibusb.c:(.text+0x4e6): undefined reference to `libusb_free_ss_usb_device_capability_descriptor'
/usr/bin/ld: testlibusb.c:(.text+0x50d): undefined reference to `libusb_free_bos_descriptor'
/usr/bin/ld: /tmp/ccL5eQxt.o: in function `print_device':
testlibusb.c:(.text+0x68e): undefined reference to `libusb_get_device_speed'
/usr/bin/ld: testlibusb.c:(.text+0x6f9): undefined reference to `libusb_get_device_descriptor'
/usr/bin/ld: testlibusb.c:(.text+0x744): undefined reference to `libusb_get_device_address'
/usr/bin/ld: testlibusb.c:(.text+0x756): undefined reference to `libusb_get_bus_number'
/usr/bin/ld: testlibusb.c:(.text+0x79e): undefined reference to `libusb_open'
/usr/bin/ld: testlibusb.c:(.text+0x7d8): undefined reference to `libusb_get_string_descriptor_ascii'
/usr/bin/ld: testlibusb.c:(.text+0x824): undefined reference to `libusb_get_string_descriptor_ascii'
/usr/bin/ld: testlibusb.c:(.text+0x87a): undefined reference to `libusb_get_string_descriptor_ascii'
/usr/bin/ld: testlibusb.c:(.text+0x8cc): undefined reference to `libusb_get_config_descriptor'
/usr/bin/ld: testlibusb.c:(.text+0x8ff): undefined reference to `libusb_free_config_descriptor'
/usr/bin/ld: testlibusb.c:(.text+0x952): undefined reference to `libusb_close'
/usr/bin/ld: /tmp/ccL5eQxt.o: in function `test_wrapped_device':
testlibusb.c:(.text+0x9d0): undefined reference to `libusb_wrap_sys_device'
/usr/bin/ld: testlibusb.c:(.text+0x9e3): undefined reference to `libusb_strerror'
/usr/bin/ld: testlibusb.c:(.text+0xa1d): undefined reference to `libusb_get_device'
/usr/bin/ld: /tmp/ccL5eQxt.o: in function `main':
testlibusb.c:(.text+0xb3c): undefined reference to `libusb_init_context'
/usr/bin/ld: testlibusb.c:(.text+0xb79): undefined reference to `libusb_get_device_list'
/usr/bin/ld: testlibusb.c:(.text+0xb8e): undefined reference to `libusb_exit'
/usr/bin/ld: testlibusb.c:(.text+0xbed): undefined reference to `libusb_free_device_list'
/usr/bin/ld: testlibusb.c:(.text+0xbf7): undefined reference to `libusb_exit'
collect2: error: ld returned 1 exit status
```

I think my distro (Fedora 40) has libusb installed as standard, and shared libraries are stored in the following locations (found using `find /usr -iname libusb*` along with various permutations of  libusbmuxd, libusbredirhost):
```
/usr/lib/libusb-1.0.so.0
/usr/lib/libusb-1.0.so.0.4.0
/usr/lib64/libusb-1.0.so.0
/usr/lib64/libusb-1.0.so.0.4.0
/usr/local/lib/pkgconfig/libusb-1.0.pc
/usr/local/lib/libusb-1.0.so.0.4.0
/usr/local/lib/libusb-1.0.so.0
/usr/local/lib/libusb-1.0.so
/usr/local/lib/libusb-1.0.la
/usr/local/lib/libusb-1.0.a
/usr/share/doc/libusb1
/usr/share/licenses/libusb1
```

It appears that the compiler isn't finding the shared library. What I'm wondering if if this seems like an issue with the compiler, or if you can see something going wrong with my install of libusb.

Thanks!

**Comment by mcuee** [comment]
Please use `pkg-config` to find the right way of compiler and linker option.

**Comment by tormodvolden** [comment, marked as answer]
> Note I do not see why I'd need to do this is there's a shared library

The compiler (or the linker) doesn't add libraries by itself. You'll have to add `-lusb-1.0` to your gcc command line to tell it to link against a libusb-1.0.so library. Since you have both system-installed versions and your own versions you must take care so that the right one is chosen for linking, and that the same one is dynamically linked when you run the program. So you might need `-L/usr/local/bin` as well, and then when running the program, make sure /usr/local/lib is in the ld.so.conf path before /usr/lib64, or set LD_LIBRARY_PATH.

pkg-config will give you the right flags with the --cflags and --libs options, but with multiple installations also here you might have to tell it with PKG_CONFIG_PATH where to read the correct configuration.

**Reply by PCBscanner** [reply]
Thanks - this worked for me. The reason I built from source because even compiling with `gcc testlibusb.c -o testlibusb -L/usr/lib -lusb-1.0`, whilst having libusb.h in the examples directory did not work despite this being where the system-installed shared libraries are located.

**Reply by PCBscanner** [reply]
Just wanted to add a bit more context after digging further. I have `/usr/lib/libusb-1.0.so.0` & `.so.0.4` installed on my system as standard, as well as the same two libraries in the `/usr/lib64` directory. Compiling with `gcc -L/usr/lib64 -Wall -o testlibusb testlibusb.c -l:libusb-1.0.so.0` was successful, `-lusb-1.0` alone was not. It looks like I need to update my symbolic links somewhere - that's something outside the scope of libusb but thought it's worth noting down in case someone else could find this useful.
